src/snmp_agent.c

```diff
@@ -0,0 +1,1513 @@
+      ERROR(PLUGIN_NAME ": No OIDs defined for '%s'", dd->name);
+      return (-EINVAL);
+    }
+
+    if (dd->is_instance) {
+      ERROR(PLUGIN_NAME
+            ": Instance flag can't be specified for scalar data '%s'",
+            dd->name);
+      return (-EINVAL);
+    }
+
+    if (!dd->type) {
+      ERROR(PLUGIN_NAME ": Type not defined for data '%s'", dd->name);
+      return (-EINVAL);
+    }
+  }
+
+  return (0);
+}
+
+static int snmp_agent_table_row_remove(table_definition_t *td,
+                                       const char *instance) {
+  int *index;
+  char *ins;
+
+  if ((c_avl_get(td->instance_index, instance, (void**) &index) != 0) ||
+      (c_avl_get(td->index_instance, index, (void**) &ins) != 0))
+    return (0);
+
+  pthread_mutex_lock(&g_agent->agentx_lock);
+
+  for (llentry_t *de = llist_head(td->columns); de != NULL; de = de->next) {
+    data_definition_t *dd = (data_definition_t *)de->value;
+
+    for (int i = 0; i < dd->oids_len; i++)
+      snmp_agent_unregister_oid_index(&dd->oids[i], *index);
+  }
+
+  snmp_agent_unregister_oid_index(&td->index_oid, *index);
+
+  pthread_mutex_unlock(&g_agent->agentx_lock);
+
+  DEBUG(PLUGIN_NAME ": Removed row for '%s' table [%d, %s].", td->name, *index,
+        ins);
+
+  notification_t n = {NOTIF_WARNING, cdtime(), "", "", PLUGIN_NAME, "", "", "",
+                      NULL};
+  sstrncpy(n.host, hostname_g, sizeof(n.host));
+  sstrncpy(n.plugin_instance, ins, sizeof(n.plugin_instance));
+  ssnprintf(n.message, sizeof(n.message),
+            "Removed data row from table %s instance %s index %d", td->name,
+            ins, *index);
+  plugin_dispatch_notification(&n);
+
+  c_avl_remove(td->index_instance, index, NULL, (void**) &ins);
+  c_avl_remove(td->instance_index, instance, NULL, (void**) &index);
+  sfree(index);
+  sfree(ins);
+
+  return (0);
+}
+
+static int snmp_agent_clear_missing(const value_list_t *vl,
+                                    __attribute__((unused)) user_data_t *ud) {
+  if (vl == NULL)
+    return (-EINVAL);
+
+  for (llentry_t *te = llist_head(g_agent->tables); te != NULL; te = te->next) {
+    table_definition_t *td = (table_definition_t *)te->value;
+
+    for (llentry_t *de = llist_head(td->columns); de != NULL; de = de->next) {
+      data_definition_t *dd = (data_definition_t *)de->value;
+
+      if (!dd->is_instance) {
+        if (CHECK_DD_TYPE(dd, vl->plugin, vl->plugin_instance, vl->type,
+                          vl->type_instance)) {
+          snmp_agent_table_row_remove(td, vl->plugin_instance);
+          return (0);
+        }
+      }
+    }
+  }
+
+  return (0);
+}
+
+static void snmp_agent_free_data(data_definition_t **dd) {
+
+  if (dd == NULL || *dd == NULL)
+    return;
+
+  /* unregister scalar type OID */
+  if ((*dd)->table == NULL) {
+    for (int i = 0; i < (*dd)->oids_len; i++)
+      unregister_mib((*dd)->oids[i].oid, (*dd)->oids[i].oid_len);
+  } else {
+    /* unregister all table OIDs */
```

Comment:
snmp_agent_free_data() frees scalar and tabular data. If data is tabular it will remove all rows of table for specified column. For example, we have unicast and broadcast packets counters as columns in the table, so if to call this  function with broadcast dd as a parameter it will free data for each interface for broadcast counter only. In turn, snmp_agent_free_table() iterates through all columns (dd) and calls snmp_agent_free_data().